plugins/login/login.js

```diff
@@ -13,37 +13,64 @@ const creds = require('./credentials');
  */
 module.exports = (config) => ({
     method: ['GET', 'POST'],
-    path: '/login',
+    path: '/login/{impersonate*}',
     config: {
         description: 'Login route',
         notes: 'Authenticate user with github oauth provider',
         tags: ['api', 'login'],
         auth: {
-            strategy: 'oauth'
+            strategy: 'oauth',
+            mode: 'try'
         },
         handler: (request, reply) => {
-            /* istanbul ignore if */
             if (!request.auth.isAuthenticated) {
                 return reply(boom.unauthorized(
                     `Authentication failed due to: ${request.auth.error.message}`
                 ));
             }
-            const username = request.auth.credentials.profile.username;
 
+            const factory = request.server.app.userFactory;
+            const githubToken = request.auth.credentials.token;
+            let username = request.auth.credentials.profile.username;
+            let scope = ['user'];
+
+            // Check whitelist
             if (config.whitelist.length > 0 && !config.whitelist.includes(username)) {
                 return reply(boom.forbidden(
                     `User ${username} is not whitelisted to use the api`
                 ));
             }
 
-            const profile = creds.generateProfile(username, ['user']);
-            const token = request.server.plugins.login.generateToken(username, ['user']);
+            // Check admin
+            if (config.admins.length > 0 && config.admins.includes(username)) {
+                scope.push('admin');
+            }
 
-            request.cookieAuth.set(profile);
+            // Check impersonate
+            if (request.params.impersonate) {
+                if (!scope.includes('admin')) {
+                    return reply(boom.forbidden(
+                        `User ${username} is not an admin and cannot impersonate`
+                    ));
+                }
+                let isBuild = false;
 
-            const githubToken = request.auth.credentials.token;
+                [username, isBuild] = request.params.impersonate.split('/');
+                scope = [
```

Comment:
We reassigned `scope` here, so that means it won't include `admin` in the `scope` anymore. Is that what we want? If that's the case, we can just use a flag to indicate it's an admin to replace line 46.